File: examples/Fortran/encode_and_decode.f90
program f_sample
use jhpcndf
implicit none
integer(8),parameter    :: num_data=10000
logical,parameter       :: is_relative=.true.
character(13),parameter :: enc="binary_search"
integer(4)              :: i, error_count, key, key2
real(4)                 :: tolerance
! FREALOPTによりr4/r8を切り替え
real, allocatable  :: random_data(:)
real, allocatable  :: work(:)
real, allocatable  :: encoded_lower(:)
real, allocatable  :: encoded(:)

! setup
tolerance=0.01
allocate(random_data(num_data))
allocate(encoded(num_data))
allocate(encoded_lower(num_data))
call random_number(random_data)



!
! JHPCN-DF encode example
!

call jhpcndf_encode(num_data, random_data, encoded, encoded_lower, tolerance, is_relative, enc)

!
! JHPCN-DF decode example
!

allocate(work(num_data))
call jhpcndf_decode(num_data, encoded, encoded_lower, work)
deallocate(encoded)
deallocate(encoded_lower)


! compare original and read data
error_count=0
do i=1, num_data
if (random_data(i) .ne. work(i)) then 
    if (error_count .lt. 10) then
        write(0,*) i,' th entry is not same:', random_data(i), work(i)
    endif
    error_count=error_count+1
endif
enddo

deallocate(work)
deallocate(random_data)

if (error_count .gt. 0) then
    write(0,*) error_count,' entry is not correctly restored.' 
else
    write(0,*) num_data,' data is correctly encode and decode.'
endif


end program

File: examples/C/encode_and_decode.c
#include <stdlib.h>
#include <stdio.h>
#include "jhpcndf.h"

#define NUM_DATA 100000

#ifdef _REAL_IS_DOUBLE_
#define REAL_TYPE double
#else
#define REAL_TYPE float
#endif


int main(int argc, char *argv[])
{

  //
  // JHPCN-DF Write example
  //

  // SetUp Data
  const float tolerance=0.01;
  const int num_data=NUM_DATA;
  REAL_TYPE* random_data= malloc(num_data*sizeof(REAL_TYPE));

  for(int i=0; i<num_data; i++)
  {
    random_data[i]=rand()/((REAL_TYPE)(RAND_MAX)+1);
  }

  REAL_TYPE* encoded= malloc(num_data*sizeof(REAL_TYPE));
  REAL_TYPE* encoded_lower= malloc(num_data*sizeof(REAL_TYPE));


  // Encode
#ifdef _REAL_IS_DOUBLE_
  JHPCNDF_encode_double(num_data, random_data,  encoded, encoded_lower, tolerance, 1, "binary_search");
#else
  JHPCNDF_encode_float(num_data, random_data,  encoded, encoded_lower, tolerance, 1, "binary_search");
#endif


  //
  // JHPCN-DF Read example
  //

  REAL_TYPE* work = malloc(num_data*sizeof(REAL_TYPE));

  // Read and decode
#ifdef _REAL_IS_DOUBLE_
  JHPCNDF_decode_double(num_data, encoded, encoded_lower, work);
#else
  JHPCNDF_decode_float(num_data, encoded, encoded_lower, work);
#endif



  // compare files
  int error_count=0;
  for(int i=0; i<num_data; i++)
  {
    if (random_data[i] != work[i])
    {
      if(error_count <10)
      {
        fprintf(stderr, "%d th data is not same.\n",i);
      }
      error_count++;
    }
  }
  if (error_count >0)
  {
    fprintf(stderr, "%d entry is not correctly restored.\n", error_count);
  }else{
    fprintf(stderr, "%d data is correctly write and read.\n", num_data);
  }

  return 0;
}
